shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java

```diff
@@ -352,45 +362,48 @@ private String getExaminationFirstDatasetUrl(Examination examination) {
 		}
 		return null;
 	}
 
-	public Attributes getDicomAttributesForStudy(Study study) throws PacsException {
-		long ts = new Date().getTime();
-		Examination exam = getFirstIfExist(study.getExaminations());
-		if (exam == null) return null;
-		Attributes result = getDicomAttributesForExamination(exam);
-		LOG.debug("get DICOM attributes for study " + study.getId() + " : " + (new Date().getTime() - ts) + " ms");
-		return result;
+	static String getFirstDatasetUrl(Dataset dataset) {
+		boolean condition = dataset != null 
+				&& dataset.getDatasetExpressions() != null 
+				&& !dataset.getDatasetExpressions().isEmpty()
+				&& dataset.getDatasetExpressions().get(0) != null 
+				&& DatasetExpressionFormat.DICOM.equals(dataset.getDatasetExpressions().get(0).getDatasetExpressionFormat())
+				&& dataset.getDatasetExpressions().get(0).getDatasetFiles() != null
+				&& !dataset.getDatasetExpressions().get(0).getDatasetFiles().isEmpty()
+				&& dataset.getDatasetExpressions().get(0).getDatasetFiles().get(0) != null;
+		if (condition) {
+			DatasetFile datasetFile = dataset.getDatasetExpressions().get(0).getDatasetFiles().get(0);
+			return StringUtils.replace(datasetFile.getPath(), "%20", " ");
+		} else {
+			return null;
+		}
 	}
 
-
-	public Attributes getDicomAttributesForExamination(Examination examination) throws PacsException {
+	public AcquisitionAttributes<Long> getDicomAttributesForAcquisition(DatasetAcquisition acquisition) throws PacsException {
 		long ts = new Date().getTime();
-		DatasetAcquisition acquisition = getFirstIfExist(examination.getDatasetAcquisitions());
-		if (acquisition == null) return null;
-		Attributes result = getDicomAttributesForAcquisition(acquisition);
+		List<Dataset> datasets = new ArrayList<>();
+		if (acquisition.getDatasets() != null) {
+			for (Dataset dataset : acquisition.getDatasets()) {
+				datasets.add(dataset);
+			}
+		}
+		AcquisitionAttributes<Long> dAcquisitionAttributes = new AcquisitionAttributes<>();
+		datasets.parallelStream().forEach(
+			dataset -> {
+				try {
+					dAcquisitionAttributes.addDatasetAttributes(dataset.getId(), getDicomAttributesForDataset(dataset));
+				} catch (PacsException e) {
+					throw new RuntimeException("could not get dicom attributes from pacs", e);
+				}
+			}
+		);
 		LOG.debug("get DICOM attributes for acquisition " + acquisition.getId() + " : " + (new Date().getTime() - ts) + " ms");
-		return result;
-	}
-
-	public Attributes getDicomAttributesForAcquisition(DatasetAcquisition acquisition) throws PacsException {
-		long ts = new Date().getTime();
-		Dataset ds = getFirstIfExist(acquisition.getDatasets());
-		if (ds == null) return null;
-		Attributes result = getDicomAttributesForDataset(ds);
-		LOG.debug("get DICOM attributes for dataset " + ds.getId() + " : " + (new Date().getTime() - ts) + " ms");
-		return result;
+		return dAcquisitionAttributes;
 	}
 
 	public String getDicomJson(Examination examination) throws IOException {
```

Comment:
Hi @julien-louis, could you please replace the stackTrace with a logger here in this method? Thank you!

Reply:
please remove the unused method getDicomJson, not referenced anywhere, if my search is not wrong

Reply:
THX